docs: rewrite resource-validation (#1336)
nginxinc/nginx-gateway-fabric: site/content/overview/resource-validation.md

The HTTPRoute "coffee" is invalid: spec.hostnames[0]: Invalid value: "cafe.!@#$%example.com": spec.hostnames[0] in body should match '^(\*\.)?[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$'
```

> {{< note >}}While unlikely, bypassing this validation step is possible if the Gateway API CRDs are modified to remove the validation.If this happens, Step 4 will reject any invalid values (from NGINX perspective).{{< /note >}}

Reviewer: ```suggestion
{{< note >}}While unlikely, bypassing this validation step is possible if the Gateway API CRDs are modified to remove the validation. If this happens, Step 4 will reject any invalid values (from NGINX perspective).{{< /note >}}
```

## Step 2 - CEL or Webhook validation by Kubernetes

- **Kubernetes 1.25 and later - CEL validation by Kubernetes API Server**

   The Kubernetes API server validates Gateway API resources using CEL validation embedded in the Gateway API CRDs. It validates Gateway API resources using advanced rules unavailable in the OpenAPI schema validation. For example, if you create a Gateway resource with a TCP listener that configures a hostname, the CEL validation will reject it with the following error:

   ```shell
   kubectl apply -f some-gateway.yaml
   ```

   ```text
   The Gateway "some-gateway" is invalid: spec.listeners: Invalid value: "array": hostname must not be specified for protocols ['TCP', 'UDP']
   ```

   More information on CEL in Kubernetes can be found [here](https://kubernetes.io/docs/reference/using-api/cel/).

- **Kubernetes 1.23 and 1.24 - Webhook validation by Gateway API Webhook**

>    The validating webhook must be installed for these Kubernetes versions {{link to installation doc}}. It validates Gateway API resources using advanced rules unavailable in the OpenAPI schema validation. For example, if you create a Gateway resource with a TCP listener that configures a hostname, the webhook will reject it with the following error:

Reviewer 2: ```suggestion
The validating webhook must be [installed for these Kubernetes versions]({{<relref "installation/installing-ngf/helm/#installing-the-gateway-api-resources">}}. It validates Gateway API resources using advanced rules unavailable in the OpenAPI schema validation. For example, if you create a Gateway resource with a TCP listener that configures a hostname, the webhook will reject it with the following error:
```

Reviewer: Why not use the doc from the includes section: https://github.com/nginxinc/nginx-gateway-fabric/blob/main/site/content/includes/installation/install-gateway-api-resources.md?

Author: Should this link be specifically to the Helm guide, to the manifests guide, either, or both?

Author: ```suggestion
   The validating webhook must be installed [with Helm]({{<relref "installation/installing-ngf/helm/#installing-the-gateway-api-resources">}} or [manifests]({{<relref "installation/installing-ngf/manifests/#installing-the-gateway-api-resources">}} for these Kubernetes versions. It validates Gateway API resources using advanced rules unavailable in the OpenAPI schema validation. For example, if you create a Gateway resource with a TCP listener that configures a hostname, the webhook will reject it with the following error:
```

This is my suggestion to provide both options. 
@ADubhlaoich your suggestion is missing 3 spaces in the beginning for the indentation.

Reviewer: The instructions for installing the gateway api resources are exactly the same in the helm and manifests docs. I'm not sure why we need both here.

Author: I will add 3 spaces to Alan's suggestion and apply that

Author: ```suggestion
   The validating webhook must be [installed for these Kubernetes versions]({{<relref "installation/installing-ngf/helm/#installing-the-gateway-api-resources">}}. It validates Gateway API resources using advanced rules unavailable in the OpenAPI schema validation. For example, if you create a Gateway resource with a TCP listener that configures a hostname, the webhook will reject it with the following error:
```

>    The Gateway API comes with a validating webhook which is enabled by default in the Gateway API installation manifests. It validates Gateway API resources using advanced rules unavailable in the OpenAPI schema validation. For example, if you create a Gateway resource with a TCP listener that configures a hostname, the webhook will reject it with the following error:

Reviewer: This is no longer true. The user has to install the webhook separately now if they are running on k8s 1.23 or 1.24.

Reviewer 2: @kate-osborn Does installing the webhook separately already have instructions elsewhere that can be linked to?

It sounds like this could probably be broken into two sections based on the version they're running: it's dependent on if the expectation is that the reader is always running the latest available versions.

Reviewer: @Jcahilltorre my suggestion did not include the actual link to the installation doc. It was a placeholder since I wasn't sure how to format the link. The instructions are here: https://github.com/nginxinc/nginx-gateway-fabric/blob/main/site/content/includes/installation/install-gateway-api-resources.md

Author: My mistake on accepting the suggestion without thorough review. Please, going forward, only use the suggestion option for ready-to-commit suggestions. If changes are required use a regular comment and call out the bits that need to be completed.

>    The validating webhook must be [installed for these Kubernetes versions]({{<relref "installation/installing-ngf/helm/#installing-the-gateway-api-resources">}}. It validates Gateway API resources using advanced rules unavailable in the OpenAPI schema validation. For example, if you create a Gateway resource with a TCP listener that configures a hostname, the webhook will reject it with the following error:

Reviewer: ```suggestion
   The validating webhook must be [installed for these Kubernetes versions]({{<relref "installation/installing-ngf/helm/#installing-the-gateway-api-resources">}}). It validates Gateway API resources using advanced rules unavailable in the OpenAPI schema validation. For example, if you create a Gateway resource with a TCP listener that configures a hostname, the webhook will reject it with the following error:
```

   ```shell
   kubectl apply -f some-gateway.yaml
   ```

   ```text
   Error from server: error when creating "some-gateway.yaml": admission webhook "validate.gateway.networking.k8s.io" denied the request: spec.listeners[1].hostname: Forbidden: should be empty for protocol TCP
   ```

{{< note >}}Bypassing this validation step is possible if the webhook is not running in the cluster. If this happens, Step 3 will reject the invalid values.{{< /note >}}

## Step 3 - Webhook validation by NGINX Gateway Fabric

> To ensure that the resources are validated with the webhook validation rules, even if the webhook is not running, NGINX Gateway Fabric performs the same validation. However, NGINX Gateway Fabric performs the validation __after_ the Kubernetes API server accepts the resource.

Reviewer: ```suggestion
To ensure that the resources are validated with the webhook validation rules, even if the webhook is not running, NGINX Gateway Fabric performs the same validation. However, NGINX Gateway Fabric performs the validation _after_ the Kubernetes API server accepts the resource.
```

Events:
  Type     Reason    Age   From                            Message
  ----     ------    ----  ----                            -------

>   Warning  Rejected  6s    nginx-gateway-fabric-nginx  the resource failed webhook validation, however the Gateway API webhook failed to reject it with the error; make sure the webhook is installed and running correctly; validation error: spec.listeners[1].hostname: Forbidden: should be empty for protocol TCP; NGF will delete any existing NGINX configuration that corresponds to the resource

Reviewer 3: This also requires a basic code change. `internal/mode/static/state/store.go` line 289 is where that NGF string is actually created.

Author: reverted this change

Reviewer 3: Why revert? Don't we want to avoid acronyms?

Author: as per our conversation, modified to NGINX Management Suite in doc and go file

>   Warning  Rejected  6s    nginx-gateway-fabric-nginx  the resource failed webhook validation, however the Gateway API webhook failed to reject it with the error; make sure the webhook is installed and running correctly; validation error: spec.listeners[1].hostname: Forbidden: should be empty for protocol TCP; NGINX Gateway Fabric will delete any existing NGINX configuration that corresponds to the resource

Author: ```suggestion
  Warning  Rejected  6s    nginx-gateway-fabric-nginx  the resource failed webhook validation, however the Gateway API webhook failed to reject it with the error; make sure the webhook is installed and running correctly; validation error: spec.listeners[1].hostname: Forbidden: should be empty for protocol TCP; NGF will delete any existing NGINX configuration that corresponds to the resource
```

>   Warning  Rejected  6s    nginx-gateway-fabric-nginx  the resource failed webhook validation, however the Gateway API webhook failed to reject it with the error; make sure the webhook is installed and running correctly; validation error: spec.listeners[1].hostname: Forbidden: should be empty for protocol TCP; NGINX Management Suite will delete any existing NGINX configuration that corresponds to the resource

Reviewer: ```suggestion
  Warning  Rejected  6s    nginx-gateway-fabric-nginx  the resource failed webhook validation, however the Gateway API webhook failed to reject it with the error; make sure the webhook is installed and running correctly; validation error: spec.listeners[1].hostname: Forbidden: should be empty for protocol TCP; NGINX Gateway Fabric will delete any existing NGINX configuration that corresponds to the resource
```

```

{{< note >}}This validation step always runs and cannot be bypassed. NGINX Gateway Fabric will ignore any resources that fail the webhook validation, like in the example above. If the resource previously existed, NGINX Gateway Fabric will remove any existing NGINX configuration for that resource.{{< /note >}}

## Step 4 - Validation by NGINX Gateway Fabric

This step catches the following cases of invalid values:

> - Valid values from the Gateway API perspective but not supported by NGINX Gateway Fabric yet. For example, a feature in an   HTTPRoute routing rule. For the list of supported features see [Gateway API Compatibility](gateway-api-compatibility.md) doc.

Reviewer 3: ```suggestion
- Valid values from the Gateway API perspective but not supported by NGINX Gateway Fabric yet. For example, a feature in an HTTPRoute routing rule. For the list of supported features see [Gateway API Compatibility](gateway-api-compatibility.md) doc.
```
